localstack/aws/handlers/cors.py

```diff
@@ -72,25 +74,34 @@
     "Access-Control-Expose-Headers",
 ]
 
-ALLOWED_CORS_ORIGINS = [
-    "https://app.localstack.cloud",
-    "http://app.localstack.cloud",
-    f"https://localhost:{config.EDGE_PORT}",
-    f"http://localhost:{config.EDGE_PORT}",
-    f"https://localhost.localstack.cloud:{config.EDGE_PORT}",
-    f"http://localhost.localstack.cloud:{config.EDGE_PORT}",
-    "https://localhost",
-    "https://localhost.localstack.cloud",
-    # for requests from Electron apps, e.g., DynamoDB NoSQL Workbench
-    "file://",
-]
-if EXTRA_CORS_ALLOWED_ORIGINS:
-    ALLOWED_CORS_ORIGINS += EXTRA_CORS_ALLOWED_ORIGINS.split(",")
 
-ACL_REQUEST_PRIVATE_NETWORK = "Access-Control-Request-Private-Network"
-ACL_ALLOW_PRIVATE_NETWORK = "Access-Control-Allow-Private-Network"
-
-LOG = logging.getLogger(__name__)
+def _get_allowed_cors_origins() -> List[str]:
+    """Construct the list of allowed origins for CORS enforcement purposes"""
+    result = [
+        # allow access from Web app and localhost domains
+        "https://app.localstack.cloud",
+        "http://app.localstack.cloud",
+        "https://localhost",
+        "https://localhost.localstack.cloud",
+        # for requests from Electron apps, e.g., DynamoDB NoSQL Workbench
+        "file://",
+    ]
+    # Add allowed origins for localhost domains, using different protocol/port combinations.
+    # If a different port is configured for EDGE_PORT_HTTP, add it to allowed origins as well
+    _ports = set([config.EDGE_PORT] + ([config.EDGE_PORT_HTTP] if config.EDGE_PORT_HTTP else []))
+    for protocol in {"http", "https"}:
+        for port in _ports:
+            result.append(f"{protocol}://localhost:{port}")
+            result.append(f"{protocol}://localhost.localstack.cloud:{port}")
```

Comment:
on question: how does `HOSTNAME_EXTERNAL` factor into this? should we consider this variable at all? i can also see we're hardcoding `localhost.localstack.cloud` instead of using `LOCALHOST_HOSTNAME`, is that intentional?

/cc @simonrw

Reply:
Good point - adding `LOCALHOST_HOSTNAME` instead of hardcoding the string makes sense! (will update..)

`HOSTNAME_EXTERNAL` is slightly different and does not directly factor into this - `HOSTNAME_EXTERNAL` would be the hostname we return in URLs that point back to LocalStack APIs (e.g., SQS queue URLs), whereas the CORS origins are Web applications (accessed via the browser) that are permitted to access our APIs. 

There might potentially be a case for either S3 websites or CloudFront distributions where we'd want to make special subdomains accessible via CORS (to be able to serve HTML/JS pages using endpoints like `http://mybucket.s3.localhost.localstack.cloud:4566`). /cc @bentsku  I'd say we can tackle those in a follow-up, once we have a better understanding of those use cases (may require regex matching on the domains, which could add a bit more complexity). 👍

Reply:
Just chiming in here: For S3, this is already implemented by using a "preprocess" handler (which performs S3 specific CORS handling before the actual CORS handler is called). This doesn't really add additional CORS origins, but handles dynamic CORS rules which can be configured on a bucket-level in S3.
However, if we find more use cases where the CORS handling has to be adjusted for a specific service (f.e. cloudfront or API gateway), we might want to implement an explicit API for services such that they can influence the CORS rules and enforcement.

Reply:
Coming back to this, sorry for being so late, working with CORS at the moment.
I think I understand better the use case now: we'd want to allow our own hosted S3 website to access other LocalStack API. Configuring S3 CORS rules would not matter, as it is the other way around (allowing some JS code hosted in our bucket to target our LocalStack instance).

Also, just watching the demo app, it doesn't seem to me that we're deploying this bucket as a S3 Website. We're just serving an HTML file, but we're not supporting any S3 website features. An S3 website would be accessible using http://<BUCKET_NAME>.s3-website.localhost.localstack.cloud:<port> only, in the same fashion as AWS.

Coming back to the use case, we could either regex match our known domains, or we could dynamically add origins to our CORS handler.